### ISSUE TITLE
Buttons Unresponsive

### ISSUE THREAD

**Issue by AryasDad** [unfoldedcircle/feature-and-bug-tracker#176]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

Hard buttons are unresponsive in an activity.  You have to exit out of the activity and re-enter for buttons to become responsive.  Happens several times.

it is frustrating when testing out certain commands for a device as it leads one to believe that the command is not correct when in reality the button is just not firing as expected.

### How to Reproduce

1. Activate Activity
2. Press buttons - no response
3. Exit activity back to home screen
4. Re-enter activity
5. Press button - works as expected

### Expected behavior

Buttons work anytime you enter an activity the 1st time, 2nd time, to nth time.

### System version

1.4.5

### What part of the system affected by the problem?

UI

**Comment by rccoleman** [comment]
This is still happening with the just-released v1.52 firmware.  It's a pretty frustrating experience.

**Comment by AryasDad** [comment]
Yes, just experienced it again after I had to force a reboot to solve an endless unresponsive error loop.  Fired up Apple TV activity after reboot and had to exit out and back in for buttons to work.

**Comment by RC-Thoughts** [comment]
I can confirm this. Slightly annoying.

**Comment by janusloo** [comment]
I encountered this behaviour as well when the Prevent Sleep option is disabled in the activity page. Just enable it for now until they resolve this in the next firmware release.

**Comment by RC-Thoughts** [comment]
> I encountered this behaviour as well when the Prevent Sleep option is disabled in the activity page. Just enable it for now until they resolve this in the next firmware release.

This does unfortunately not solve the issue for me. Same issue is with Profile-pages, one has to re-activate profile after exiting activity to be able to use up/down/ok-buttons.

**Comment by janusloo** [comment]
> > This does unfortunately not solve the issue for me. Same issue is with Profile-pages, one has to re-activate profile after exiting activity to be able to use up/down/ok-buttons.

Hmm now you mentioned it... the hard buttons stopped working indeed after exiting the active activity. I used the OLED screen to swipe to another screen and back and the buttons worked again.

**Comment by pkscout** [comment]
Can confirm that this is still happening with the 1.6.1 firmware.  The "solution" for now is to only have one activity and never put the remote to sleep.  Makes for garbage user experience and horrendous battery life (I can maybe get 10 hours and then it takes 10 hours to charge).

If I reboot the remote, I have to start the activity, then exit the activity, then re-enter.  Otherwise the hard buttons still don't work.

**Comment by Quinterly** [comment]
And it continues to happen with 1.6.4 ... this is the single most important bug to fix for usability and it is getting zero attention from UC.

**Comment by zehnm** [comment]
Unfortunately we are unable to reproduce this issue and need more information to track this down. We have found and fixed button issues in 1.6.0 when leaving an activity and certain command related issues, but we couldn't reproduce any button issues within an activity anymore.

Please provide more information, either screen shots or descriptions:
- What kind of entities and integrations are used within the activity
- Button mappings of the activity
- If the activity is within an activity group
- If multiple activities are running
- Log information

Log files can be downloaded in the web-configurator under _Settings: Development, Logs_

<img width="821" alt="image" src="https://github.com/unfoldedcircle/feature-and-bug-tracker/assets/17356629/09910a05-bdd8-4814-a8a9-219518c1d0f2">

Then select:
- Download logs from: `Core service`, `User interface app`, and optionally the integrations you are using in the activity.
- Priority: DEBUG
- If the event to observe didn't happen right before downloading the log file:
  - increase the number of log entries to 1000
  - optionally set a timeframe

<img width="535" alt="image" src="https://github.com/unfoldedcircle/feature-and-bug-tracker/assets/17356629/fd06bd5d-ecf2-4f66-b958-b2018ef0260f">

**Comment by RC-Thoughts** [comment]
My issues disappeared along with , big thanks!

**Comment by Quinterly** [comment]
Mine haven't disappeared. To use Apple TV (it also happens on my other R2 with Android TV but the logs are the one dealing with Apple TV), I have to reboot the R2. Every time I want to switch to an Apple TV or Android TV integration (as long as the remote doesn't go to sleep I can flip back and forth). 

This is happening with 2 different R2s in two different setups. 

Activities are in a activity group. Buttons are mapped to cursor control, home, back, and to volume/mute of the AVR. Equipment used in the activity:

- AppleTV 4K
- Sony Bravia TV (controlled via TCP through Home Assistant)
- NAD T758v3 (controlled by HA integration and Telnet through HA). 

It is not specific to the equipment as the other remote is connected to an LG TV and a Yamaha AVR. Same issues. 
[UCR2_logs_2024-01-21(3).txt](https://github.com/unfoldedcircle/feature-and-bug-tracker/files/14002052/UCR2_logs_2024-01-21.3.txt)

**Comment by pvmil** [comment]
Haven't seen this issue anymore since last firmware and the activity setting to keep the remote from sleeping (mainly using sending commands trough home assistant)

**Comment by Quinterly** [comment]
> Haven't seen this issue anymore since last firmware and the activity setting to keep the remote from sleeping (mainly using sending commands trough home assistant)

Problem seems to be when the remote sleeps (eg. put back on the dock). Once rebooted, it works fine until the next sleep.

**Comment by kennymc-c** [comment]
This doesn't even happen for me with activities that do not prevent sleep mode. Of course I have to wait a little bit to let the remote to connect to the wifi, dock and external integrations but I do not need to exit the activity to get all buttons to work.

**Comment by zehnm** [comment]
> It is not specific to the equipment as the other remote is connected to an LG TV and a Yamaha AVR. Same issues.
> [UCR2_logs_2024-01-21(3).txt](https://github.com/unfoldedcircle/feature-and-bug-tracker/files/14002052/UCR2_logs_2024-01-21.3.txt)

@Quinterly  The logs show that the Home Assistant integration can't connect to the HA server between 15:26:34 until 16:02:12 (first "connection timeouts" then "Connection refused). That is pretty much for the whole duration of the log file. Since you wrote you control the TV and AVR through HA, they won't be controllable until the integration is connected.

The log file is missing the Apple TV integration, or are you also controlling the ATV through HA?

To me it seems that this issue is from non-connected integrations. 

Most of the errors are timeouts, the integration can't connect to the Home Assistant server within 3 seconds. I suggest increasing the connection timeout to 6 seconds to see if that improves reconnecting to HA:

1. Open the Home Assistant integration in the web configurator and click on "Start integration setup".
2. Verify server URL
3. Token can be left empty, the previous token will be used
4. Select `Configure enhanced options`
5. Next
6. Increase the `Connection timeout` value
7. Save settings with Next

<img width="534" alt="image" src="https://github.com/unfoldedcircle/feature-and-bug-tracker/assets/17356629/5b55a9eb-40eb-46ee-9fd9-2d7dc3be408e">

Another thing to try is to use an IP address, instead of the `homeassistant.local` local name. Some routers and access points are unfortunately not mDNS friendly.

**Comment by Quinterly** [comment]
@zehnm 

I am not using HA for the AppleTV integration. The issue is the UC AppleTV integration failing on two different R2s requiring a reboot to make it work. Every time ... once I exit AppleTV, I have to reboot the R2 to use the integration again. 

Here are the logs just for AppleTV. This has nothing to do with Home Assistant (which is being used for IP control of my AVR and TV). However, the advice on the HA integration errors is appreciated and will address a different issue I am having (mostly with the mute toggle not being responsive). 



[UCR2_logs_2024-01-29(2).txt](https://github.com/unfoldedcircle/feature-and-bug-tracker/files/14088937/UCR2_logs_2024-01-29.2.txt)

**Comment by Quinterly** [comment]
@zehnm Multiple firmware revisions and this is still not fixed (or even acknowledged). This, https://github.com/unfoldedcircle/feature-and-bug-tracker/issues/298 and this, https://github.com/unfoldedcircle/feature-and-bug-tracker/issues/237, are clearly related issues that you continue to ignore. It's been two weeks of radio silence since I posted a response to your feedback with logs. I know UC is small and this is a kickstarter project. But, please fix bugs that make the remote unusable before focusing on roadmap features. Simply put, I can't ask my family to use a remote that needs to be rebooted every time they want to use AppleTV or that requires them to bounce in and out of activities to get hard buttons to work.

**Comment by albaintor** [comment]
HI,

EDIT : the recent firmwares updates seem to have fixed this issue on Home assistant integration, but not on the other integrations (shield and apple TV)
I was able to reproduce the problem again : running an activity with Apple TV integration
The volumes buttons (assigned to Sony AVR through Home assistant) work
However the Apple TV buttons don't. I had to reboot the RC2 to make them work again
Here are the logs :
[UCR2_logs_2024-02-11.zip](https://github.com/unfoldedcircle/feature-and-bug-tracker/files/14231790/UCR2_logs_2024-02-11.zip)

I pressed left key 3 times without success, see this part :
```2024-02-11 08:21:48.686677 +00:00	ui	INFO	uc.ui.input: Key released: "DPAD_LEFT" "Activity_QMLTYPE_241(0x5572e6ce20)"
2024-02-11 08:21:48.685803 +00:00	ui	INFO	qml: OVERRIDE RELEASED: DPAD_LEFT Activity_QMLTYPE_241(0x5572e6ce20)
2024-02-11 08:21:48.580813 +00:00	intg-appletv	INFO	DEBUG:ucapi.api:->: {"kind": "resp", "req_id": 8, "code": 200, "msg": "result", "msg_data": {}}
2024-02-11 08:21:48.530954 +00:00	intg-appletv	INFO	DEBUG:ucapi.api:<-: {"kind":"req","id":8,"msg":"entity_command","msg_data":{"cmd_id":"cursor_left","entity_id":"F0:B3:EC:12:20:20","entity_type":"media_player"}}
2024-02-11 08:21:48.525486 +00:00	ui	INFO	uc.ui.input: Key pressed: "DPAD_LEFT" "Activity_QMLTYPE_241(0x5572e6ce20)"
2024-02-11 08:21:48.521844 +00:00	ui	INFO	qml: OVERRIDE PRESSED: DPAD_LEFT Activity_QMLTYPE_241(0x5572e6ce20)
2024-02-11 08:21:47.729217 +00:00	intg-appletv	INFO	DEBUG:ucapi.api:->: {"kind": "resp", "req_id": 7, "code": 200, "msg": "result", "msg_data": {}}
2024-02-11 08:21:47.187353 +00:00	ui	INFO	uc.ui.input: Key released: "DPAD_LEFT" "Activity_QMLTYPE_241(0x5572e6ce20)"
2024-02-11 08:21:47.186220 +00:00	ui	INFO	qml: OVERRIDE RELEASED: DPAD_LEFT Activity_QMLTYPE_241(0x5572e6ce20)
2024-02-11 08:21:47.060482 +00:00	intg-appletv	INFO	DEBUG:ucapi.api:<-: 
```


A few lines after : there is a timeout error
```
2024-02-11 08:20:06.285382 +00:00	ui	INFO	uc.ui.input: Key pressed: "POWER" "Activity_QMLTYPE_241(0x55c2070f70)"
2024-02-11 08:20:06.283793 +00:00	ui	INFO	qml: DEFAULT PRESSED: POWER QQuickContentItem(0x55c0a851b0, "ApplicationWindow")
2024-02-11 08:20:06.274735 +00:00	systemd-logind	INFO	Power key pressed short.
2024-02-11 08:20:04.435165 +00:00	ui	INFO	uc.ui.input: Key released: "DPAD_LEFT" "Activity_QMLTYPE_241(0x55c2070f70)"
2024-02-11 08:20:04.434249 +00:00	ui	INFO	qml: OVERRIDE RELEASED: DPAD_LEFT Activity_QMLTYPE_241(0x55c2070f70)
2024-02-11 08:20:04.287981 +00:00	ui	INFO	uc.ui.input: Key pressed: "DPAD_LEFT" "Activity_QMLTYPE_241(0x55c2070f70)"
2024-02-11 08:20:04.286426 +00:00	ui	INFO	qml: OVERRIDE PRESSED: DPAD_LEFT Activity_QMLTYPE_241(0x55c2070f70)
2024-02-11 08:20:02.178239 +00:00	ui	INFO	uc.ui.input: Key released: "DPAD_LEFT" "Activity_QMLTYPE_241(0x55c2070f70)"
2024-02-11 08:20:02.177370 +00:00	ui	INFO	qml: OVERRIDE RELEASED: DPAD_LEFT Activity_QMLTYPE_241(0x55c2070f70)
2024-02-11 08:20:02.018066 +00:00	ui	INFO	uc.ui.input: Key pressed: "DPAD_LEFT" "Activity_QMLTYPE_241(0x55c2070f70)"
2024-02-11 08:20:02.016620 +00:00	ui	INFO	qml: OVERRIDE PRESSED: DPAD_LEFT Activity_QMLTYPE_241(0x55c2070f70)
2024-02-11 08:20:01.587403 +00:00	ui	INFO	uc.ui.input: Key released: "DPAD_LEFT" "Activity_QMLTYPE_241(0x55c2070f70)"
2024-02-11 08:20:01.584055 +00:00	ui	INFO	qml: OVERRIDE RELEASED: DPAD_LEFT Activity_QMLTYPE_241(0x55c2070f70)
2024-02-11 08:20:01.446977 +00:00	ui	INFO	uc.ui.input: Key pressed: "DPAD_LEFT" "Activity_QMLTYPE_241(0x55c2070f70)"
2024-02-11 08:20:01.446213 +00:00	ui	INFO	qml: OVERRIDE PRESSED: DPAD_LEFT Activity_QMLTYPE_241(0x55c2070f70)
**2024-02-11 08:20:01.057987 +00:00	ui	WARN	uc.ui.entity.controller: Cannot execute command: "media_player.previous" 408 "The request timed out. Please try again.**"
```

**Comment by Yasser** [comment]
I am having the same issue on 1.6.10. I launch an activity and let the remote sleep. Upon waking, the hardware buttons don't work (some, but not all, oddly). Exiting the activity (not turning it off) and then re-launching the activity (which is already "on") re-enables the unresponsive buttons.

**Comment by martonborzak** [comment]
> I am having the same issue on 1.6.10. I launch an activity and let the remote sleep. Upon waking, the hardware buttons don't work (some, but not all, oddly). Exiting the activity (not turning it off) and then re-launching the activity (which is already "on") re-enables the unresponsive buttons.

I have been not able to reproduce this with an activity including an Apple TV and network controlled speakers. If you don't have "Prevent sleep" enabled the remote will go to sleep after the configured timeout. Depending on your network and configuration, it might take a second or so for everything to reconnect. While reconnecting, buttons assigned to those devices can seem to be not working. Exiting the activity ui and re-opening it again could be enough time for things to reconnect. Please provide logs for the core, ui and integrations involved so we can check exactly what's going on.

**Comment by albaintor** [comment]
It has been a while since I did not have this issue again. Seems to be random and not frequent so difficult to reproduce. 
Otherwise @martonborzak in the optimizations do you plan to make something as buffering buttons press (as logi harmony does) while reconnecting instead of raising error ?

Thanks

**Comment by Yasser** [comment]
I've definitely had instances where the buttons (tied to HA actions) remain unresponsive even after restarting the activity, so I'm not certain if it's purely network related or not. I'll collect the logs the next time I run into the issue.

I do agree that better error handling/logging and/or asynchronous retries for certain network errors would improve the experience.

**Comment by Yasser** [comment]
To follow up, this happened again last evening. I'm integrating with the AppleTV through HASS. 

I launched by "Watch AppleTV" activity and all hardware buttons related to the AppleTV were unresponsive. Logs from Remote Two show multiple errors i.e.: `Entity hass.main.script.atv_menu is not loaded. Button mapping failed for release: DPAD_DOWN`

I waited several minutes, and the buttons remained unresponsive, so it's not simply a matter of the remote waking from sleep. As before, exiting and relaunching the activity returned the expected behavior to the buttons.

I can post logs if helpful.

**Comment by hprotter** [comment]
I still have this issue after every restart of the remote. Hardware button in an activity will not be responsive when an activity is opened for the first time and I have to leave the activity and restart it again. After restarting the activity, all buttons will work normal. This issue is still existent in 1.7.14 firmware and this issue is existent since I own the remote from last year on.

**Comment by alexthegreat100** [comment]
This issue STILL exists in V. 2.6.12!. First time Activity in opened, buttons are unresponsive. Go back to Home screen and open the activity again. Now the buttons work. Using AppleTV-Integration.
Pls fix this!!!

**Comment by ItsHarper** [comment]
Reported as fixed in 2.7

**Comment by janusloo** [comment]
Yes I can confirm that it is now fixed in the beta version 2.7. I couldn't believe it at first but on multiple reboots of the R3 it is solid, this issue is now fixed for good! Kudos to the devs ;-)

**Comment by zehnm** [comment]
Confirmed fixed by the reporter in beta 2.7. Closing.
